pkg/authorization/rulevalidation/compact_rules.go

```diff
@@ -0,0 +1,62 @@
+package rulevalidation
+
+import (
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+)
+
+// compactRules combines rules that contain a single APIGroup/Resource, differ only by verb, and contain no other attributes.
+// this is a fast check, and works well with the decomposed "missing rules" list from a Covers check.
+func compactRules(rules []authorizationapi.PolicyRule) []authorizationapi.PolicyRule {
+	compacted := make([]authorizationapi.PolicyRule, 0, len(rules))
+
+	simpleRules := map[unversioned.GroupResource]*authorizationapi.PolicyRule{}
+	for _, rule := range rules {
+		if resource, isSimple := isSimpleResourceRule(&rule); isSimple {
+			if existingRule, ok := simpleRules[resource]; ok {
+				// Add the new verbs to the existing simple resource rule
+				existingRule.Verbs.Insert(rule.Verbs.List()...)
+			} else {
+				// Copy the rule and the verbs string set
```

Comment:
I'd rather deep copy the whole thing.  This returns a franken object where some fields of the return are shared with input and some are not, right?